alf/algorithms/lagrangian_algorithm.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) 2021 Horizon Robotics and ALF Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Lagrangian Algorithm."""
+
+import numpy as np
+import gin
+
+import torch
+import torch.nn as nn
+from torch.nn import functional as F
+
+import alf
+from alf.algorithms.algorithm import Algorithm
+from alf.data_structures import namedtuple, AlgStep
+from alf.nest import nest
+from alf.tensor_specs import TensorSpec
+from alf.utils import tensor_utils
+
+LagInfo = namedtuple("LagInfo", ["rollout_reward"], default_value=())
+
+
+@alf.configurable(blacklist=["reward_spec"])
+class LagrangianAlgorithm(Algorithm):
+    """An algorithm that adjusts reward weights according to untransformed
+    rollout rewards. The adjustment is expected to be performed after every
+    training iteration.
+
+    Generally speaking, for each reward dimension, the algorithm compares the
+    expected undiscounted episodic return to a predefined threshold, and if the
+    return is greater than the threshold (requirement satisfied) then it decreases
+    the weight; otherwise it increases the weight.
+
+    Usually the threshold is defined for a fixed time length, so equivalently,
+    for simplicity, the algorithm can instead compares individual rewards per
+    step to an averaged threshold.
```

Comment:
To make it less confusing, you don't need to mention "expected undiscounted episodic return". Just the desired threshold for reward per step.

Reply:
I think the reason why I mention "expected undiscounted episodic return" here is that this is exactly what this algorithm is doing, as long as we only have a single lambda for each reward dim. We are using the expectation of immediate rewards (equal to the return with a scale) to tune that single lambda, and the desired threshold is actually not set for  every step, but for an expectation. 

Another complete different way is still comparing immediate reward against threshold per step, but having a different lambda conditioned on the state. In this case, we are really putting a constraint on every step.

Reply:
What I means is that you only need to talk about the desired expected immediate reward, not about episodic.

Reply:
updated